File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.py
import sys
input = lambda: sys.stdin.readline().rstrip()
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)

def main():
    t = int(input())
    for _ in range(t):
        n = int(input())
        s = [0]+[*map(int, input().split())]
        deg = [0]*(n+1)
        for i in range(1, n+1): deg[s[i]] += 1
        st = [i for i in range(1, n+1) if deg[i] == 0]
        ans = 0
        while st:
            cur = st.pop() 
            ans += 1
            nxt = s[cur]
            deg[nxt] -= 1
            if deg[nxt] == 0: st.append(nxt)
        print(ans)  
    return
if __name__ == "__main__":
    main()

File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.cc
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int t; cin >> t;
    while(t--){
        int n; cin >> n;
        vector<int> s(n+1, 0);
        for(int i = 1; i <= n; i++) cin >> s[i];
        vector<int> deg(n+1, 0);
        for(int i = 1; i <= n; i++) deg[s[i]]++;
        stack<int> st;
        for(int i = 1; i <= n; i++) if(deg[i] == 0) st.push(i);
        int ans = 0;
        while(st.size()){
            int cur = st.top(); st.pop();
            ans++;
            int nxt = s[cur];
            deg[nxt]--;
            if(deg[nxt] == 0) st.push(nxt);
        }
        cout << ans << '\n';
    }
    return 0;
}